### ISSUE TITLE
question in model training

### ISSUE THREAD

**Issue by kaishxu** [Georgetown-IR-Lab/epic-neural-ir#1]
Hello,

I am replicating the EPIC method. My doubt is that is this model fully trained?
I notice that the model stops learning at around 60 steps with the batch size equal to 16. However, no matter which dataset you use, it seems only a small part of the dataset is used in trainning.

PS: Because of poor coding ability, I feel so hard to read your code. Could you please tell me which dataset do you use (triples.train.small.tar.gz)? How could you sample the negative passages? What is your loss function?

**Comment by kaishxu** [comment]
well, it may be a stupid question. :)
How could you load the parameters into the model? I don't notice any "epic" related variables loaded into the model. 😂

**Comment by seanmacavaney** [comment]
Hi Kaishuai,

I do not think that looking at the percentage of data used for fine-tuning is all that useful in an IR setting to determine whether a model is undertrained. The reason being that in a dataset like MS-MARCO, there are roughly 4.7e12 possible training triples (# of positive judgments in MS-MARACO train x number of documents). My understanding is that the training triples file that the dataset authors provide is meant to allow for more meaningful comparisons between systems by providing a standard (though not compulsory) training sequence. Besides that, there's nothing special about that file-- it just samples negatives from the top 1000 indri QL results, I believe.

What we observe when training is a _somewhat_ steady increase on the dev set, and then 20 training iterations with no improvement, so we stop there (early stopping). I agree that it could potentially be undertrained, but 20 iterations with no improvement are quite a few (often people use just 2 or 3, right?) Each training iteration consists of 512 triples in total (batch_size x batches_per_epoch), which isn't a ton, but also not a trivial amount of data, especially when you consider the number of iterations. This is the same setting that can train effective "Vanilla" BERT models. If you want to investigate this further, you can set `pipeline.early_stop=100` or something else larger like that. I'd be interested to hear what you find!

Regarding model parameters- those are query_salience (\theta_1 in the paper), doc_salience (\theta_3), and doc_quality (\teta_4). They are defined in the code [here](https://github.com/Georgetown-IR-Lab/OpenNIR/blob/master/onir/rankers/epic.py#L33) and can be [loaded in from a pytroch checkpoint](https://pytorch.org/tutorials/beginner/saving_loading_models.html) like any other parameters. All the rest of EPIC's parameters come from BERT (including \theta_2, which is BERT's cls.predictions).

The other questions I believe are answered in the paper: we use the full "official" MS-MARCO training sequence. The link to that file was broken at some point, but was fixed [May of last year](https://github.com/microsoft/MSMARCO-Passage-Ranking/commit/27c1fab833868b1c56751ce94bd49e87f19a0465). There has since been a version "2" of that file, and I'm [not really sure what the deal is with that](https://github.com/microsoft/MSMARCO-Passage-Ranking/commit/4695a71c6c76ce85c07a51c0f12690cab19abbb0). Since we're the training triples defined by the dataset, we do not do any negative sampling ourselves. We use a cross-entropy loss function.

Does this help?

 \- sean

**Comment by kaishxu** [comment]
Hi SEAN,

Thanks a lot for your reply. It helps a lot!!!!!!!
So, 
1. the training samples are directly and randomly selected from the "triples" file;
2. the loss for each sample is actually calculated by softmax and cross-entropy;
Right?

As for the number of training samples, I may try to explore further, since I notice a great difference in other papers within a similar structure (such as [ANCE](https://github.com/microsoft/ANCE), [ColBERT](https://github.com/stanford-futuredata/ColBERT), [Rep-BERT](https://github.com/jingtaozhan/RepBERT-Index), and [CLEAR](https://arxiv.org/abs/2004.13969)). They all directly use the whole training samples in the MSMARCO "triple.small" file. I hope more training samples are unessential, and it will save a lot of time to validate a model.

**Comment by seanmacavaney** [comment]
Yes, but with the slight caveat that the training samples are selected from the "full" triples file in sequence (not randomly).

Glad this helped!
